source/common/network/filter_manager_impl.h

```diff
@@ -116,21 +120,18 @@ class FilterManagerImpl : protected Logger::Loggable<Logger::Id::connection> {
   bool startUpstreamSecureTransport();
   void maybeClose();
   void onConnectionClose(ConnectionCloseAction close_action);
-
-  void finalizeClose(ConnectionCloseAction close_action) {
-    state_.pending_local_close_ = false;
-    state_.pending_remote_close_ = false;
-    connection_.closeConnection(close_action);
-  }
-
-  bool pendingClose() { return state_.pending_local_close_ || state_.pending_remote_close_; }
+  bool pendingClose() { return state_.local_close_pending_ || state_.remote_close_pending_; }
 
 protected:
   struct State {
-    uint32_t pending_close_write_filter_{0};
-    uint32_t pending_close_read_filter_{0};
-    bool pending_remote_close_{false};
-    bool pending_local_close_{false};
+    // Number of pending filters awaiting closure.
+    uint32_t filter_pending_close_count_{0};
```

Comment:
should we assert this is 0 during dtor if the runtime flag is enabled?

Reply:
no, I think there are some cases that the filter_pending_close status are bypassed, e.g., remote early close, other abort close, and so on.